rustfst/src/algorithms/minimize.rs

```diff
@@ -331,57 +331,56 @@ impl AcyclicMinimizer {
+        let mut pairs = Vec::new();
+        let mut it_partition = Vec::new();
+        let height = state_cmp.partition.num_classes();
         for h in 0..height {
-            // We need here a binary search tree in order to order the states id and create a partition.
+            // We need here a search tree with a custom comparator in order to order the states
+            // id and create a partition.
             // For now uses the crate `stable_bst` which is quite old but seems to do the job
             // TODO: Bench the performances of the implementation. Maybe re-write it.
             let mut equiv_classes =
                 TreeMap::<StateId, StateId, _>::with_comparator(|a: &StateId, b: &StateId| {
                     state_cmp.compare(*a, *b).unwrap()
                 });
 
-            let it_partition: Vec<_> = self.partition.iter(h).collect();
+            it_partition.extend(state_cmp.partition.iter(h));
             equiv_classes.insert(it_partition[0] as StateId, h as StateId);
-
-            let mut classes_to_add = vec![];
-            for e in it_partition.iter().skip(1) {
-                // TODO: Remove double lookup
-                if equiv_classes.contains_key(&(*e as StateId)) {
-                    equiv_classes.insert(*e as StateId, NO_STATE_ID);
-                } else {
-                    classes_to_add.push(e);
-                    equiv_classes.insert(*e as StateId, NO_STATE_ID);
-                }
-            }
-
-            for v in classes_to_add {
-                equiv_classes.insert(*v as StateId, self.partition.add_class() as StateId);
+            classes_to_add.clear();
+            for &e in it_partition.iter().skip(1) {
+                equiv_classes.get_or_insert(e as StateId, || {
+                    classes_to_add.insert(e);
+                    NO_STATE_ID
+                });
             }
-
-            for s in it_partition {
-                let old_class = self.partition.get_class_id(s);
-                let new_class = *equiv_classes.get(&(s as StateId)).unwrap();
-                if new_class == NO_STATE_ID {
+            pairs.extend(it_partition.drain(..).map(|s| {
+                (s, *equiv_classes.get(&(s as StateId)).unwrap())
+            }));
+            for (s, c) in pairs.drain(..) {
+                let old_class = state_cmp.partition.get_class_id(s);
+                let new_class = if classes_to_add.contains(&s) {
+                    debug_assert_eq!(c, NO_STATE_ID);
+                    state_cmp.partition.add_class()
+                } else if c != NO_STATE_ID {
+                    c
+                } else {
                     // The behaviour here is a bit different compared to the c++ because here
                     // when inserting an equivalent key it modifies the key
                     // which is not the case in c++.
                     continue;
-                }
-                if old_class != (new_class as usize) {
```

Comment:
If I remember correctly, this cast was necessary because of the `state-label-u32` feature. Maybe that's no longer the case. Did you manage to run the tests ? It seems the CI hasn't been run 🤔

Reply:
I ran the tests, but not with all features enabled. Fixing this...

Reply:
This should be fixed with my latest commit.